core/runtime/jsruntime.rs

```diff
@@ -550,6 +552,78 @@ pub struct RuntimeOptions {
   /// through evalContext.
   pub eval_context_code_cache_cbs:
     Option<(EvalContextGetCodeCacheCb, EvalContextCodeCacheReadyCb)>,
+
+  pub import_assertions_support: ImportAssertionsSupport,
+}
+
+#[derive(Default)]
+pub enum ImportAssertionsSupport {
+  // TODO(bartlomieju): we probably don't want it to be default, but rather `Warning` or `Error`.
+  // `assert` keyword is supported.
+  #[default]
+  Yes,
+  // `assert` keyword is supported, but prints a warning message on occurence.
+  Warning,
+  // `assert` keyword is no longer supported and causes a SyntaxError
+  Error,
+  // `assert` keyword is supported, provided callback is called on each occurence.
+  CustomCallback(Box<dyn Fn()>),
+}
+
+impl ImportAssertionsSupport {
+  fn is_enabled(&self) -> bool {
+    matches!(self, Self::Yes | Self::Warning | Self::CustomCallback(_))
+  }
+
+  fn has_warning(&self) -> bool {
+    matches!(self, Self::Warning | Self::CustomCallback(_))
+  }
+}
+
+/// Currently only handles warnings for "import assertions" deprecation
+extern "C" fn isolate_message_listener(
+  message: v8::Local<v8::Message>,
+  _exception: v8::Local<v8::Value>,
+) {
+  eprintln!("isoalte message listener");
+  let scope = &mut unsafe { v8::CallbackScope::new(message) };
+  let scope = &mut v8::HandleScope::new(scope);
+  let message_v8_str = message.get(scope);
+  let message_str = message_v8_str.to_rust_string_lossy(scope);
+
+  if !message_str.starts_with("'assert' is deprecated") {
+    return;
+  }
+
+  let maybe_script_resource_name = message.get_script_resource_name(scope);
+  let maybe_source_line = message.get_source_line(scope);
+  let maybe_line_number = message.get_line_number(scope);
+  let start_column = message.get_start_column();
+
+  let js_runtime_state = JsRuntime::state_from(scope);
+  match &js_runtime_state.import_assertions_support {
+    ImportAssertionsSupport::Warning => {
+      eprintln!("⚠️  Import assertions are deprecated. Use `with` keyword, instead of 'assert' keyword.");
+      if let Some(specifier) = maybe_script_resource_name {
+        if let Some(source_line) = maybe_source_line {
+          eprintln!();
+          eprintln!("{}", source_line.to_rust_string_lossy(scope));
+          eprintln!();
+          eprintln!("{:0width$}^^^^^^", " ", width = start_column);
+        }
+        eprintln!(
+          "  at {}:{}:{}",
+          specifier.to_rust_string_lossy(scope),
+          maybe_line_number.unwrap(),
+          start_column
+        );
+      }
+    }
```

Comment:
This prints a message like:
```
⚠️  Import assertions are deprecated. Use `with` keyword, instead of 'assert' keyword.

import data from "./data.json" assert { type: "json" };

                               ^^^^^^
  at file:///Users/ib/dev/deno/foo.js:1:31
```

Kinda feel we might want to reuse it in the CLI, just print it in yellow (with a `Warning:` prefix). Need to figure out how to easily share it